### ISSUE TITLE
linter: `react/button-has-type` reports false positive when assigning type dynamically

### ISSUE THREAD

**Issue by alex3683** [oxc-project/oxc#20546]
### What version of Oxlint are you using?

1.56.0

### What command did you run?

`npm exec oxlint -- -c=./.oxlintrc.json`

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
{
  "$schema": "./node_modules/oxlint/configuration_schema.json",
  "plugins": ["import", "react", "typescript", "unicorn"],
  "categories": {
    "correctness": "warn",
    "pedantic": "allow", // "warn" would be nice, but currently gives about a 1000 warnings ...
    "restriction": "warn",
    "suspicious": "warn"
  },
  "env": {
    "browser": true,
    "es2020": true
  },
  "ignorePatterns": [
    "**/build/**",
    "**/dist/**",
    "**/node_modules/**",
    "**/public/**",
    "**/temp/**",
    "**/tsconfig.tsbuildinfo",
    "**/vite.config.*",
    "custom.d.ts",
    "eslint.config.mjs"
  ],
  "options": {
    "typeAware": true,
    "typeCheck": true
  },
  "settings": {
    "react": {
      "version": "17.0"
    }
  },
  "rules": {
    // nothing in regard to react/button-has-type
  }
}

```

### What happened?

The type for the button can be configured as property, but is limited to valid values only by the typing. Nonetheless oxlint reports a violation here, although dynamic type assignment should be allowed according to the original implementation: https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/button-has-type.md

This is my reduced example:

```ts
export interface TransparentButtonProps {
  children?: ReactNode;
  type?: 'button' | 'submit' | 'reset';
  onClick: () => void;
}

export const TransparentButton = observer((props: TransparentButtonProps): ReactNode => {
  return (
    <button
      type={props.type ?? 'button'} // <-- This is reported as false positive
      onClick={props.onClick}
    >
      {props.children}
    </button>
  );
});
```

Message:

```
`button` elements must have a valid `type` attribute.
help: Change the `type` attribute to one of the allowed values: `button`, `submit`, or `reset`.oxc[eslint-plugin-react(button-has-type)](https://oxc.rs/docs/guide/usage/linter/rules/react/button-has-type.html)
```

**Comment by camc314** [comment]
> although dynamic type assignment should be allowed according to the original implementation: https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/button-has-type.md

@alex3683 could you explain this in more detail?

the docs in the upstream version of this rule say:

> Examples of incorrect code for this rule:
> ```
> var Hello = <button>Hello</button>
> var Hello = <button type="foo">Hello</button>
> var Hello = <button type={foo}>Hello</button>
> ```

The last example is a dynamic assignment, and they report an error on that code.

**Comment by alex3683** [comment]
@camc314 Fair enough. I was referring to `var Hello = <button type={condition ? "button" : "submit"}>Hello</button>` as a correct code example, which in my opinion is also dynamic. I guess the thing is, that in untyped javascript the variable `foo` could be anything, while in a small AST as for the ternary operator only the two options have to be checked.

However when working with typescript this seems arbitrary since in my example from above its easy to reason what possible values of the variable could be by looking at the type. So yeah, maybe this issue could be interpreted as a feature request to transform this rule to being type aware and notice, whether not only actual values meet the criteria of the rule, but also all values possible by the type definition.

**Comment by mehm8128** [comment]
Although the original rule reports assignment values as invalid, I think we can consider to allow them.
Since other rules which check for enum values, such as `jsx-a11y/autocomplete-valid` or `jsx-a11y/lang`, allow dynamic values, we could treat `react/button-has-type` in the same way.

[playground](https://playground.oxc.rs/?lintRules=jsx-a11y%2Fautocomplete-valid%2Cjsx-a11y%2Flang%2Creact%2Fbutton-has-type&options=%7B%22run%22%3A%7B%22lint%22%3Atrue%2C%22formatter%22%3Afalse%2C%22transform%22%3Afalse%2C%22isolatedDeclarations%22%3Afalse%2C%22whitespace%22%3Afalse%2C%22mangle%22%3Afalse%2C%22compress%22%3Afalse%2C%22scope%22%3Atrue%2C%22symbol%22%3Atrue%2C%22cfg%22%3Afalse%7D%2C%22parser%22%3A%7B%22extension%22%3A%22tsx%22%2C%22allowReturnOutsideFunction%22%3Atrue%2C%22preserveParens%22%3Atrue%2C%22allowV8Intrinsics%22%3Atrue%2C%22semanticErrors%22%3Atrue%7D%2C%22linter%22%3A%7B%7D%2C%22formatter%22%3A%7B%22useTabs%22%3Afalse%2C%22tabWidth%22%3A2%2C%22endOfLine%22%3A%22lf%22%2C%22printWidth%22%3A80%2C%22singleQuote%22%3Afalse%2C%22jsxSingleQuote%22%3Afalse%2C%22quoteProps%22%3A%22as-needed%22%2C%22trailingComma%22%3A%22all%22%2C%22semi%22%3Atrue%2C%22arrowParens%22%3A%22always%22%2C%22bracketSpacing%22%3Atrue%2C%22bracketSameLine%22%3Afalse%2C%22objectWrap%22%3A%22preserve%22%2C%22singleAttributePerLine%22%3Afalse%7D%2C%22transformer%22%3A%7B%22target%22%3A%22es2015%22%2C%22useDefineForClassFields%22%3Atrue%2C%22experimentalDecorators%22%3Atrue%2C%22emitDecoratorMetadata%22%3Atrue%2C%22optimizeEnums%22%3Atrue%2C%22optimizeConstEnums%22%3Atrue%7D%2C%22isolatedDeclarations%22%3A%7B%22stripInternal%22%3Afalse%7D%2C%22codegen%22%3A%7B%22normal%22%3Atrue%2C%22jsdoc%22%3Atrue%2C%22annotation%22%3Atrue%2C%22legal%22%3Atrue%7D%2C%22compress%22%3A%7B%7D%2C%22mangle%22%3A%7B%22topLevel%22%3Atrue%2C%22keepNames%22%3Afalse%7D%2C%22controlFlow%22%3A%7B%22verbose%22%3Afalse%7D%2C%22inject%22%3A%7B%22inject%22%3A%7B%7D%7D%2C%22define%22%3A%7B%22define%22%3A%7B%7D%7D%7D&code=%3C%3E%0A%3Chtml+lang%3D%22invalid-lang%22%3E%3C%2Fhtml%3E%0A%3Chtml+lang%3D%22ja%22%3E%3C%2Fhtml%3E%0A%3Chtml+lang%3D%7Bvalue%7D%3E%3C%2Fhtml%3E%0A%0A%3Cinput+autocomplete%3D%22invalid-value%22+%2F%3E%0A%3Cinput+autocomplete%3D%22name%22+%2F%3E%0A%3Cinput+autocomplete%3D%7Bvalue%7D+%2F%3E%0A%0A%3Cbutton+type%3D%22invalid-type%22%3E%3C%2Fbutton%3E%0A%3Cbutton+type%3D%22button%22%3E%3C%2Fbutton%3E%0A%3Cbutton+type%3D%7Bvalue%7D%3E%3C%2Fbutton%3E%0A%3C%2F%3E)

**Comment by jz-cp** [comment]
Adding my +1 to this. Running into this as well.
